TestModels/Union/runtimes/rust/src/types/_my_union.rs

```diff
@@ -0,0 +1,51 @@
+// Code generated by software.amazon.smithy.rust.codegen.smithy-rs. DO NOT EDIT.
+#[allow(missing_docs)] // documentation missing in model
+#[non_exhaustive]
+#[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::std::fmt::Debug)]
+pub enum MyUnion {
+    #[allow(missing_docs)] // documentation missing in model
+    IntegerValue(i32),
+    #[allow(missing_docs)] // documentation missing in model
+    StringValue(::std::string::String),
+    /// The `Unknown` variant represents cases where new union variant was received. Consider upgrading the SDK to the latest available version.
+    /// An unknown enum variant
+    ///
+    /// _Note: If you encounter this error, consider upgrading your SDK to the latest version._
+    /// The `Unknown` variant represents cases where the server sent a value that wasn't recognized
+    /// by the client. This can happen when the server adds new functionality, but the client has not been updated.
+    /// To investigate this, consider turning on debug logging to print the raw HTTP response.
+    #[non_exhaustive]
+    Unknown,
+}
+impl MyUnion {
+    /// Tries to convert the enum instance into [`IntegerValue`](crate::types::MyUnion::IntegerValue), extracting the inner [`i32`](i32).
+    /// Returns `Err(&Self)` if it can't be converted.
+    pub fn as_integer_value(&self) -> ::std::result::Result<&i32, &Self> {
+        if let MyUnion::IntegerValue(val) = &self {
+            ::std::result::Result::Ok(val)
+        } else {
+            ::std::result::Result::Err(self)
+        }
+    }
+    /// Returns true if this is a [`IntegerValue`](crate::types::MyUnion::IntegerValue).
+    pub fn is_integer_value(&self) -> bool {
+        self.as_integer_value().is_ok()
```

Comment:
Out of curiousity, is this as efficient as something like:

```rust
if let MyUnion::IntegerValue(val) = &self {
    true
} else {
    false
}
```

If this is what smithy-rs generates then I'm sure it meets the right bar mind you.